diff --git a/dwm/config.def.h b/dwm/config.def.h
index ee88065..6aec5cf 100644
--- a/dwm/config.def.h
+++ b/dwm/config.def.h
@@ -104,6 +104,8 @@ static const Layout layouts[] = {
 /* helper for spawning shell commands in the pre dwm-5.0 fashion */
 #define SHCMD(cmd) { .v = (const char*[]){ "/bin/sh", "-c", cmd, NULL } }
 
+#include "movestack.c"
+
 /* commands */
 static char dmenumon[2] = "0"; /* component of dmenucmd, manipulated in spawn() */
 static const char *dmenucmd[] = { "dmenu_run", "-m", dmenumon, "-fn", dmenufont, "-nb", col_gray1, "-nf", col_gray3, "-sb", col_cyan, "-sf", col_gray4, NULL };
@@ -153,6 +155,10 @@ static const Key keys[] = {
 	{ MODKEY,			XK_j,	   focusstack,	   {.i = +1 } },
 	{ MODKEY,			XK_k,	   focusstack,	   {.i = -1 } },
 
+	// Move windows in stack
+	{ MODKEY|ShiftMask,		XK_j,	   movestack,	   {.i = +1 } },
+	{ MODKEY|ShiftMask,		XK_k,	   movestack,	   {.i = -1 } },
+
 	// Increase/decrease number of master windows
 	{ MODKEY,			XK_i,	   incnmaster,	   {.i = +1 } },
 	{ MODKEY,			XK_d,	   incnmaster,	   {.i = -1 } },
